### PR TITLE
fix(index): `tapable` deprecation warnings (`webpack >= v4.0.0`)

### PR DESCRIPTION
**Pull request by rynclark**
### `Notable Changes`

Adds support for `webpack >= 4`.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -30,65 +82,39 @@ export default class BabelMinifyPlugin {
...
-          .filter(ModuleFilenameHelpers.matchObject.bind(null, options))
-          .forEach((file) => {
-            try {
-              const asset = compilation.assets[file];
-
-              if (asset.__babelminified) {
-                compilation.assets[file] = asset.__babelminified;
-                return;
-              }
+    const isWebpack4 = !!compiler.hooks;
```
Comment (michael-ciniawsky) [inline src/index.js]: Since this check is only used once `if (compiler.hooks)` is sufficient

File: src/index.js [hunk 2]
```diff
@@ -30,65 +82,39 @@ export default class BabelMinifyPlugin {
...
 
-              let input;
-              let inputSourceMap;
+    if (isWebpack4) {
+      compiler.hooks
```
Comment (michael-ciniawsky) [inline src/index.js]:
```js
if (compiler.hooks)
  const { plugin } = this
  const { compilation } = compiler.hooks
  
  compilation.tap(plugin, ...)
  ...
}

File: src/index.js [hunk 3]
```diff
@@ -30,65 +82,39 @@ export default class BabelMinifyPlugin {
...
+        .compilation
+        .tap(NAMESPACE, (compilation) => {
```
Comment (michael-ciniawsky) [inline src/index.js]: The compilation callbacks could also be extracted into a `{Function}` 

```js
function buildModuleFn () {}
function optimizeChunkAssetsFn () {}

function compilationFn (compilation) {
   if (compilation.hooks) {
      const { plugin } = this
      const { buildModule, optimizeChunksAssets } = compilation.hooks
      
      buildModule.tap(plugin, buildModuleFn)
      optiimizeChunksAssets.tap(plugin, optimizeChunkAssetsFn)
   } else {
      compilation.plugin('build-module', buildModuleFn)
      compilation.plugin('optimize-chunk-assets', optimizeChunkAssetsFn)
   }
}

if (compiler.hooks) {
   ...
   compilation.tap(plugin, compilationFn)
} else {
   compiler.plugin('compilation', compilationFn)
}
```

File: src/index.js [hunk 4]
```diff
@@ -27,68 +112,17 @@ export default class BabelMinifyPlugin {
...
-              compilation.errors.push(e);
-            }
-          });
+    if (compiler.hooks) {
+      const { compilation } = compiler.hooks;
 
-        callback();
-      });
-    });
+      compilation.tap(this.plugin, compilation => compilationFn(compilation, this));
```
Comment (michael-ciniawsky) [inline src/index.js]:
```js
compilation.tap(this.plugin, compilationFn.bind(this, compilation));
```

File: src/index.js [hunk 5]
```diff
@@ -27,68 +112,17 @@ export default class BabelMinifyPlugin {
...
+    } else {
+      compiler.plugin('compilation', compilation => compilationFn(compilation, this));
```
Comment (michael-ciniawsky) [inline src/index.js]:
```js
 compiler.plugin('compilation', compilationFn.bind(this, compilation));
```

File: package.json [hunk 6]
```diff
@@ -29,10 +29,11 @@
     "travis:coverage": "npm run test:coverage -- --runInBand",
     "travis:lint": "npm run lint && npm run security",
     "travis:test": "npm run test -- --runInBand",
-    "webpack-defaults": "webpack-defaults"
+    "webpack-defaults": "webpack-defaults",
```
Comment (michael-ciniawsky) [inline package.json]:
```diff
- "webpack-defaults": "webpack-defaults",
```

File: package.json [hunk 7]
```diff
@@ -28,9 +28,7 @@
     "test:watch": "jest --watch",
     "travis:coverage": "npm run test:coverage -- --runInBand",
     "travis:lint": "npm run lint && npm run security",
-    "travis:test": "npm run test -- --runInBand",
-    "webpack-defaults": "webpack-defaults",
-    "defaults": "webpack-defaults"
```
Comment (michael-ciniawsky) [inline package.json]: `"defaults": "webpack-defaults"`  should stay :P

File: src/index.js [hunk 8]
```diff
@@ -8,8 +8,90 @@ function getDefault(actualValue, defaultValue) {
...
+      } catch (e) {
+        compilation.errors.push(e);
+      }
+    });
+}
+
+function compilationFn(useHooks, compilation) {
+  const { options, plugin } = this;
+
+  if (useHooks) {
```
Comment (michael-ciniawsky) [inline src/index.js]: `useHooks` => `if (compilation.hooks)`

**Comment by michael-ciniawsky** [comment]
Release blocked by https://status.npmjs.org/incidents/gbk5bjx1169j

File: src/index.js [hunk 9]
```diff
@@ -30,65 +112,13 @@ export default class BabelMinifyPlugin {
...
-
-              asset.__babelminified = compilation.assets[file] = result.map
-                ? new SourceMapSource(result.code, file, result.map, input, inputSourceMap)
-                : new RawSource(result.code);
-            } catch (e) {
-              compilation.errors.push(e);
-            }
-          });
+    if (compiler.hooks) {
+      const { compilation } = compiler.hooks;
```
Comment (ooflorent) [inline src/index.js]: In webpack we banned this pattern because it prevents searching for `hooks.*`.

Comment (michael-ciniawsky) [reply src/index.js]: hmm... 🤔 debatable imho :P, personally _always_ in favor of compacter code and here searching for the `hookName` (e.g `compilation`) instead then. A search for `hooks*`, one still gets where the hooks got destructured. Maybe destructuring is more confusing then helpful within larger parts of the `webpack` code base e.g `Parser` or the like ¯\_(ツ)_/¯. Or did you mean something entirely different, with can break here ? 🙃

Anyways if there is general consensus on avoiding to destructure `.hooks` fine with either... :)

**Comment by michael-ciniawsky** [comment]
Released in [`v0.3.1`](https://github.com/webpack-contrib/babel-minify-webpack-plugin/releases/tag/v0.3.1) 🎉 Thx

**Pull request merged**